### DISCUSSION TITLE
KH10 limit battery charge current

### DISCUSSION THREAD

**Discussion by cperricos** [nathanmarlor/foxess_modbus#595] [Q&A]
Hello,
I have a new Fox ess solar KH10 installation and love the modbus integration. I would like ideally to manage the loads on my supply (I need to keep it under 80A), and so ideally would want to limit the charge current to the batter when I top up the battery at night  as it can pull up to 10KW. Is there a way of doing this via HA?

many thanks

**Comment by FozzieUK** [comment]
Yes that is possible.

You would do it by controlling the maximum charge current (it’s DC, to the charge power multiply it by your battery volts)

Probably easiest to have a simple automation that sets it to a lower value at night for when you charge from grid, and then reset it to 40A (the default) for day time charging from PV.

**Comment by cperricos** [comment]
That sounds perfect -  please forgive my ignorance but how do I set the maximum charge current from home assistant. I was unsure what sensor to set.
thanks again

**Reply by FozzieUK** [reply]
The sensor to modify is:-    number.max_charge_current

You can set it like this in an automation action as follows -

![IMG_1576](https://github.com/nathanmarlor/foxess_modbus/assets/113460294/e3913b03-8910-4572-87e0-63a7a7eac4b7)

**Reply by cperricos** [reply]
So I tried that last night and unfortunately that didnt seem to work. As soon as the battery charge kicked in, the current went up to nearly 40 Amps. This is the code I used in the automsation (which Ran with no issues):

id: 'set_charge_current_to_10'
alias: "Set Charge Current to 10"
description: "Set charge current to 10 between midnight and 10 AM"
trigger:
platform: time
at: "00:00:00"
action:
service: input_number.set_value
target:
entity_id: number.foxess_max_charge_current
data:
value: 10
mode: single
Any suggestions would be appreciated. Thanks a gain for all your help

**Reply by FozzieUK** [reply]
Aaah yes I spotted it straight away, it’s definitely a subtle one

There are 2 types of number set service call, one for number helpers (input_number.set_value) and one for number entities (number.set_value).

You have used the input_number.set which won’t work here as it is a number entity - you’ll need to change this line

```service: input_number.set_value```

to 

```service: number.set_value```

You can test it to make sure it is working, when you are in the automation, the 3 dots to the right hand side of each action drops down a menu and you can ‘run’ each action independently from there - if you open another tab and watch the value of ```number.foxess_max_charge_current``` you should see it change.

**Reply by cperricos** [reply]
So I tried this and its not working either. The register. I have also tried to display number.foxess_max_charge_current and it is showing as unavailable on the dashboard. The logs did throw up this curious error though:

Logger: custom_components.foxess_modbus.modbus_controller
Source: custom_components/foxess_modbus/modbus_controller.py:315
integration: FoxESS - Modbus ([documentation](https://github.com/nathanmarlor/foxess_modbus), [issues](https://github.com/nathanmarlor/foxess_modbus/issues))
First occurred: 4:18:58 PM (1 occurrences)
Last logged: 4:18:58 PM

tcp://192.168.8.60:502 247 - 5 failed poll attempts: now not connected. Last error: Error reading registers. Type: RegisterType.HOLDING; start: 31000; count: 55; slave: 247 from tcp://192.168.8.60:502: Exception Response(131, 3, IllegalAddress)

could this be a problem with the inverter? I havent upgraded the firmware since I got it as I read on this forum that there were issues with later ones:

Master version: 1.18
Slave version: 1.01
Manager Version: 1.13

many thanks in advance in your help

**Reply by FozzieUK** [reply]
Ok yes agree, I think that’s the problem here, you need to be on the latest version of KH firmware and also the latest version of this integration. 

The latest versions for the KH are now - 
Master: 1.30
Manager: 1.22

These are much more stable than the inter-mediate firmware releases.

Can you arrange to get them updated at your convenience, then once you’ve done that make sure that you are on the latest beta version of the modbus integration (as of today it’s v1.10.2b5)

**Reply by cperricos** [reply]
Update should be done in next 48 hours. Also sorry but is there an obvious way to force an upgrade to the latest beta? have looked all over the integration in HA and cant find a way to do this. Sorry if I have missed something obvious

**Reply by FozzieUK** [reply]
Great, hopefully not long to wait 👍

On the beta versions probably easiest if I link you to this description [hacs beta](https://hacs.xyz/docs/faq/beta/)

**Comment by cperricos** [comment]
That is brilliant - thank you!
  
  
  
  
  
  
  

  
  
>

**Reply by cperricos** [reply]
So I tried that last night and unfortunately that didnt seem to work. As soon as the battery charge kicked in, the current went up to nearly 40 Amps. This is the code I used in the automsation (which Ran with no issues):

- id: 'set_charge_current_to_10'
  alias: "Set Charge Current to 10"
  description: "Set charge current to 10 between midnight and 10 AM"
  trigger:
    - platform: time
      at: "00:00:00"
  action:
    - service: input_number.set_value
      target:
        entity_id: number.foxess_max_charge_current
      data:
        value: 10
  mode: single


Any suggestions would be appreciated. Thanks a gain for all your help

**Comment by cperricos** [comment]
Thanks Dave - that is amazing support. You are a true gent.
  
  
  
  
  
  
  

  
  
>

**Comment by cperricos** [comment]
ok perfect - thank you. I cant believe how quickly you respond. this is really great.
________________________________

**Comment by cperricos** [comment]
Thanks a lot Dave. The latest beta appears to have worked! Still getting some connection errors but will check once the firmware has been upgraded.
  

  
Thanks
  
Costi   
  
  
  
  
  
  
  

  
  
>
